src/wms/src/main/java/org/geoserver/wms/icons/IconProperties.java

```diff
@@ -130,54 +132,69 @@ public Double getHeading() {
             }
 
             @Override
-            public String href(String baseURL, String workspace, String styleName) {
+            public String href(String baseURL, WorkspaceInfo workspace, String styleName) {
 
                 try {
                     URL target = new URL(url);
                     String graphicProtocol = target.getProtocol();
 
                     if ("file".equals(graphicProtocol)) {
                         File file = URLs.urlToFile(target);
-                        File styles = null;
-                        File graphicFile = null;
-
+                        File styles;
+                        File workspaceStyles = null;
+                        File graphicFile;
                         if (file.isAbsolute()) {
                             GeoServerDataDirectory dataDir =
                                     (GeoServerDataDirectory) GeoServerExtensions.bean("dataDirectory");
                             // we grab the canonical path to make sure we can compare them, no
                             // relative parts in them and so on
                             styles = dataDir.getStyles().dir().getCanonicalFile();
+                            if (workspace != null) {
+                                workspaceStyles =
+                                        dataDir.getStyles(workspace).dir().getCanonicalFile();
+                            }
                             file = graphicFile = file.getCanonicalFile();
                             if (file.getAbsolutePath().startsWith(styles.getAbsolutePath())) {
                                 // ok, part of the styles directory, extract only the relative path
                                 String relativePath = file.getAbsolutePath()
                                         .substring(styles.getAbsolutePath().length() + 1);
                                 file = new File(relativePath);
-                            } else {
-                                // we wont' transform this, other dirs are not published
-                                file = null;
-                            }
-
-                            // rebuild the icon href accordingly
-                            if (file != null && styles != null) {
-                                return ResponseUtils.buildURL(
-                                        baseURL,
-                                        "styles/" + styles.toURI().relativize(graphicFile.toURI()),
-                                        null,
-                                        URLType.RESOURCE);
-                            } else {
-                                // we don't know how to handle this then...
-                                return null;
+                                if (file != null) {
```

Comment:
And here: I don't think file can ever be null here

Reply:
it does not hurt

Reply:
Looking closer you are right, the constructor will never return null. Furthermore the `relativePath` is not useful since it's recomputed later with `relativize()`, so I have removed these couple of lines